core/DataAccess/LogQueryBuilder/JoinGenerator.php

```diff
@@ -0,0 +1,264 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\DataAccess\LogQueryBuilder;
+
+use Exception;
+use Piwik\Common;
+use Piwik\Tracker\LogTable;
+
+class JoinGenerator
+{
+    /**
+     * @var JoinTables
+     */
+    protected $tables;
+
+    /**
+     * @var bool
+     */
+    private $joinWithSubSelect = false;
+
+    /**
+     * @var string
+     */
+    private $joinString = '';
+
+    /**
+     * @var array
+     */
+    private $nonVisitJoins = array();
+
+    public function __construct(JoinTables $tables)
+    {
+        $this->tables = $tables;
+        $this->addMissingTablesNeededForJoins();
+    }
+    
+    private function addMissingTablesNeededForJoins()
+    {
+        foreach ($this->tables as $index => $table) {
+            if (is_array($table)) {
+                continue;
+            }
+
+            $logTable = $this->tables->getLogTable($table);
+
+            if (!$logTable->getColumnToJoinOnIdVisit()) {
+                $tableNameToJoin = $logTable->getLinkTableToBeAbleToJoinOnVisit();
+
+                if ($index > 0 && !$this->tables->hasJoinedTable($tableNameToJoin)) {
+                    $this->tables->addTableToJoin($tableNameToJoin);
+                }
+
+                if ($this->tables->hasJoinedTable($tableNameToJoin)) {
+                    $this->generateNonVisitJoins($table, $tableNameToJoin, $index);
+                }
+            }
+        }
+    }
+
+    /**
+     * Generate the join sql based on the needed tables
+     * @throws Exception if tables can't be joined
+     * @return array
+     */
+    public function generate()
+    {
+        /** @var LogTable[] $availableLogTables */
+        $availableLogTables = array();
+
+        $this->tables->sort(array($this, 'sortTablesForJoin'));
```

Comment:
we sort tables so they will be sorted in similar order as before. This seemed to be important